lib/graphql/schema/validator/required_validator.rb

```diff
@@ -51,27 +51,34 @@ def initialize(one_of: nil, argument: nil, message: nil, **default_options)
         end
 
         def validate(_object, context, value)
-          matched_conditions = 0
+          fully_matched_conditions = 0
+          partially_matched_conditions = 0
 
           if !value.nil?
             @one_of.each do |one_of_condition|
               case one_of_condition
               when Symbol
                 if value.key?(one_of_condition)
-                  matched_conditions += 1
+                  fully_matched_conditions += 1
                 end
               when Array
-                if one_of_condition.all? { |k| value.key?(k) }
-                  matched_conditions += 1
-                  break
+                full_match = one_of_condition.all? { |k| value.key?(k) }
+                partial_match = !full_match && one_of_condition.any? { |k| value.key?(k) }
```

Comment:
I discovered that `#some?` is only available in Rails 🙃

Reply:
I think we could do it in a single iteration of `one_of_condition`: 

```suggestion
                any_match = false 
                full_match = true
                one_of_condition.each do |k|
                  if value.key?(k)
                    any_match = true
                  else 
                    full_match = false 
                  end 
                end 
                
                partial_match = any_match && !full_match 
```

Reply:
Thanks for this suggestion, I'll go ahead and implement that!

Is this suggestion motivated by potential performance concerns? My GraphQL performance knowledge is limited. For most cases I imagine that one loop versus two loops for a relatively small amount of options would have a negligible impact. Of course though, it is still additional CPU cycles. But this is where my expertise and use case awareness falls off - are there use cases where this could potentially cause a performance impact?

Reply:
Yes, it was performance-minded. In practice, I can't think of a situation where this would be called in a tight loop and become a real bottleneck. But I have certainly been surprised _before_ about how rubber meets the road when GraphQL-Ruby is used in real-world applications. 

In micro-benchmarks, it makes a difference, for example: 

```ruby
# loops.rb
require "benchmark/ips"


one_of_condition = [:a, :b, :c, :d]
value = {
  a: 1,
  b: 2,
  c: 3
}
Benchmark.ips do |x|
  x.report(".all + .any") do
    full_match = one_of_condition.all? { |k| value.key?(k) }
    partial_match = !full_match && one_of_condition.any? { |k| value.key?(k) }
  end
  x.report(".each") do
    any_match = false
    full_match = true
    one_of_condition.each do |k|
      if value.key?(k)
        any_match = true
      else
        full_match = false
      end
    end

    partial_match = any_match && !full_match
  end

  x.compare!
end
```

Without YJIT, `.all` + `.any` is 23% slower: 

```
$ ruby loops.rb
ruby 3.4.1 (2024-12-25 revision 48d4efcb85) +PRISM [x86_64-darwin22]
Warming up --------------------------------------
         .all + .any   247.839k i/100ms
               .each   273.355k i/100ms
Calculating -------------------------------------
         .all + .any      2.269M (± 5.2%) i/s  (440.78 ns/i) -     11.401M in   5.039404s
               .each      2.797M (± 4.3%) i/s  (357.52 ns/i) -     14.214M in   5.091697s

Comparison:
               .each:  2797017.5 i/s
         .all + .any:  2268712.9 i/s - 1.23x  slower
```

Interestingly, with YJIT enabled, there's a much bigger difference (greater than 2x faster with `.each`): 

```
$ ruby --yjit loops.rb
ruby 3.4.1 (2024-12-25 revision 48d4efcb85) +YJIT +PRISM [x86_64-darwin22]
Warming up --------------------------------------
         .all + .any   324.644k i/100ms
               .each   759.224k i/100ms
Calculating -------------------------------------
         .all + .any      3.406M (± 5.4%) i/s  (293.57 ns/i) -     17.206M in   5.066546s
               .each      8.336M (± 5.4%) i/s  (119.95 ns/i) -     41.757M in   5.025170s

Comparison:
               .each:  8336485.6 i/s
         .all + .any:  3406394.2 i/s - 2.45x  slower
```

Maybe YJIT has special handling for `Array#each` 🤷

Reply:
I really appreciate you taking the time to explain this, and provide this fulsome example. There is clearly a difference here. While it is minimal in this context, it's easy to address and removes any potential for this to be an added source of slowness.